### PR TITLE
[MI-2748]: Implemented 1:1 mapping between MM user and Azure DevOps user for oAuth.

### PR DESCRIPTION
**Pull request by avas27JTG**
Implemented 1:1 mapping between MM user and Azure DevOps user for oAuth.

### CODE REVIEW

File: server/constants/store.go [hunk 1]
```diff
@@ -10,9 +10,10 @@ const (
...
-	ProjectKey         = "%s_%s"
-	ProjectPrefix      = "project_list"
-	SubscriptionPrefix = "subscription_list"
-	UserIDPrefix       = "oAuth"
+	OAuthPrefix           = "oAuth_%s"
+	ProjectKey            = "%s_%s"
+	ProjectPrefix         = "project_list"
+	SubscriptionPrefix    = "subscription_list"
+	UserIDPrefix          = "oAuth"
+	AzureDevOpsUserPrefix = "azd_userid_%s"
```
Comment (ayusht2810) [inline server/constants/store.go:18]: `azd_userId_%s` or `azd_userID_%s` maybe?

File: server/constants/oauth_config.go [hunk 2]
```diff
@@ -13,5 +13,6 @@ const (
 	// Paths
 	PathAuth = "/oauth2/authorize"
 	// #nosec G101 -- This is a false positive
-	PathToken = "/oauth2/token"
+	PathToken       = "/oauth2/token"
+	PathUserProfile = "/_apis/profile/profiles/%s"
```
Comment (ayusht2810) [inline server/constants/oauth_config.go:17]: I think this should be in routes.go as this is an AZD API path

File: server/plugin/client.go [hunk 3]
```diff
@@ -54,6 +55,17 @@ func (c *client) GenerateOAuthToken(encodedFormValues url.Values) (*serializers.
 	return oAuthSuccessResponse, statusCode, nil
 }
 
+func (c *client) GetUserProfile(id, accessToken string) (*serializers.UserProfile, int, error) {
```
Comment (ayusht2810) [inline server/plugin/client.go:58]:
```suggestion
func (c *client) GetUserProfile(ID, accessToken string) (*serializers.UserProfile, int, error) {
```

File: server/constants/oauth_config.go (hunk 2 above)
Comment (avas27JTG) [reply server/constants/oauth_config.go:17]: this is an AZD path but it is used for oAuth purpose only also it uses the base URL of oAuth which is different from the rest

**Review by ayusht2810 — APPROVED** [review]

File: server/constants/messages.go [hunk 4]
```diff
@@ -3,22 +3,23 @@ package constants
 const (
 	// TODO: all these messages are to be verified from Mike at the end
```
Comment (Kshitij-Katiyar) [inline server/constants/messages.go:4]: Is this TODO comment still required ?

File: server/constants/messages.go [hunk 5]
```diff
@@ -3,22 +3,23 @@ package constants
...
+	CreatedTask                                    = "Work item [#%d: \"%s\"](%s) of type \"%s\" was successfully created by %s."
+	TaskTitle                                      = "[%s #%d: %s](%s)"
+	PullRequestTitle                               = "[#%d: %s](%s)"
+	BuildDetailsTitle                              = "[#%s](%s): %s"
+	PipelineDetailsTitle                           = "[%s](%s): %s"
+	AlreadyLinkedProject                           = "This project is already linked."
+	NoProjectLinked                                = "No project is linked, please link a project."
+	PipelinesRequestBeingProcessed                 = "Your approval/rejection request is being processed."
+	PipelinesRequestProcessed                      = "Your approval/rejection request is processed."
+	ErrorMessageAzureDevopsAccountAlreadyConnected = "azure devops account for %s is already connected"
```
Comment (Kshitij-Katiyar) [inline server/constants/messages.go:22]: I think this should be with the other error messages in the below group of const variables

File: server/plugin/oAuth.go [hunk 6]
```diff
@@ -108,6 +108,12 @@ func (p *Plugin) OAuthComplete(w http.ResponseWriter, r *http.Request) {
 	}
 
 	if err := p.GenerateOAuthToken(code, state); err != nil {
+		if strings.Contains(err.Error(), "already connected") {
```
Comment (Kshitij-Katiyar) [inline server/plugin/oAuth.go:111]: constants for this maybe ?

File: server/plugin/oAuth.go [hunk 7]
```diff
@@ -190,6 +196,30 @@ func (p *Plugin) GenerateAndStoreOAuthToken(mattermostUserID string, oauthTokenF
 		return errors.Wrap(err, "failed to generate oAuth token")
 	}
 
+	userProfile, _, err := p.Client.GetUserProfile("me", successResponse.AccessToken)
```
Comment (Kshitij-Katiyar) [inline server/plugin/oAuth.go:199]: constant ?

File: server/plugin/oAuth.go [hunk 8]
```diff
@@ -190,6 +196,30 @@ func (p *Plugin) GenerateAndStoreOAuthToken(mattermostUserID string, oauthTokenF
...
+		}
+		return errors.Wrap(err, "failed to fetch user profile")
+	}
+
+	azureDevopsUser, err := p.Store.LoadAzureDevopsUserDetails(userProfile.ID)
+	if err != nil {
+		if _, DMErr := p.DM(mattermostUserID, constants.GenericErrorMessage, false); DMErr != nil {
+			return DMErr
+		}
+		return errors.Wrap(err, "failed to get user details")
```
Comment (Kshitij-Katiyar) [inline server/plugin/oAuth.go:212]:
```suggestion
		return errors.Wrap(err, "failed to get the user details")
```

File: server/store/utils.go [hunk 9]
```diff
@@ -132,6 +132,10 @@ func GetOAuthKey(mattermostUserID string) string {
 	return fmt.Sprintf(constants.OAuthPrefix, mattermostUserID)
 }
 
+func GetAzureDevOpsUserKey(azureDevopsUserID string) string {
```
Comment (Kshitij-Katiyar) [inline server/store/utils.go:135]:
```suggestion
func GetAzureDevopsUserKey(azureDevopsUserID string) string {
```

File: server/plugin/client.go (hunk 3 above)
Comment (avas27JTG) [reply server/plugin/client.go:58]: @ayusht2810 
image

File: server/plugin/oAuth.go (hunk 6 above)
Comment (avas27JTG) [reply server/plugin/oAuth.go:111]: I don't think that will be of more use as this is the only place where we need this string also the string itself is descriptive so creating a constant for readability won't be that needy

File: server/constants/oauth_config.go [hunk 10]
```diff
@@ -13,5 +13,8 @@ const (
...
+
+	CurrentAzureDevopsUserProfileId = "me"
```
Comment (Kshitij-Katiyar) [inline server/constants/oauth_config.go:19]:
```suggestion
	CurrentAzureDevopsUserProfileID = "me"
```

**Review by Kshitij-Katiyar — APPROVED** [review]

**Pull request merged**